File: XMLCheck/Middleware/ApiKeyMiddleware.cs
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Configuration;
using System.Text.Json;
using System.Threading.Tasks;

namespace XMLCheck.Middleware
{
    public class ApiKeyMiddleware
    {
        private readonly RequestDelegate _next;
        private const string APIKEYHEADER = "X-API-Key";

        public ApiKeyMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task InvokeAsync(HttpContext context, IConfiguration configuration)
        {
            var path = context.Request.Path.Value;

            // Bỏ qua kiểm tra cho các route /swagger và /health
            if (path != null && (path.StartsWith("/swagger", System.StringComparison.OrdinalIgnoreCase) || 
                                 path.StartsWith("/health", System.StringComparison.OrdinalIgnoreCase)))
            {
                await _next(context);
                return;
            }

            if (!context.Request.Headers.TryGetValue(APIKEYHEADER, out var extractedApiKey))
            {
                context.Response.StatusCode = 401;
                context.Response.ContentType = "application/json";
                var response = new { success = false, message = "Invalid API Key" };
                await context.Response.WriteAsync(JsonSerializer.Serialize(response));
                return;
            }

            var apiKey = configuration.GetValue<string>("ApiSettings:ApiKey");

            if (string.IsNullOrEmpty(apiKey) || !apiKey.Equals(extractedApiKey))
            {
                context.Response.StatusCode = 401;
                context.Response.ContentType = "application/json";
                var response = new { success = false, message = "Invalid API Key" };
                await context.Response.WriteAsync(JsonSerializer.Serialize(response));
                return;
            }

            await _next(context);
        }
    }
}

File: XMLValidator_v2/src/middleware/api-key.middleware.ts
import { Injectable, NestMiddleware } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { appConfig } from '../config/app.config';

@Injectable()
export class ApiKeyMiddleware implements NestMiddleware {
  use(req: Request, res: Response, next: NextFunction) {
    const path = req.path;

    // Bypass API key check for Swagger and health paths
    if (
      path &&
      (path.toLowerCase().startsWith('/swagger') ||
        path.toLowerCase().startsWith('/health'))
    ) {
      return next();
    }

    const API_KEY_HEADER = 'x-api-key'; // express headers are lowercase
    const extractedApiKey = req.headers[API_KEY_HEADER];

    const config = appConfig();
    const apiKey = config.apiKey;

    if (!extractedApiKey || extractedApiKey !== apiKey) {
      return res.status(401).json({
        success: false,
        message: 'Invalid API Key',
      });
    }

    next();
  }
}
